# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/streams/ReadableStream.h
# Commit: 19547ae8dc19
# Full Hash: 19547ae8dc19492947d8b2100fc1df793f2fce6e
# Author: Tom Schuster <evilpies@gmail.com>
# Date: 2022-12-21 21:21:23
# Regressor Bug: 1734244
# File Overlap Count: 1
# Description:
#   Bug 1734244 - Implement async iteration of ReadableStream. r=mgaudet,peterv
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D153312
# ==============================================================================

diff -r a186000b3a89 -r 19547ae8dc19 dom/streams/ReadableStream.h
--- a/dom/streams/ReadableStream.h	Tue Dec 20 16:11:53 2022 +0000
+++ b/dom/streams/ReadableStream.h	Tue Dec 20 16:26:25 2022 +0000
@@ -12,6 +12,7 @@
 #include "mozilla/Attributes.h"
 #include "mozilla/ErrorResult.h"
 #include "mozilla/dom/BindingDeclarations.h"
+#include "mozilla/dom/IterableIterator.h"
 #include "mozilla/dom/QueuingStrategyBinding.h"
 #include "mozilla/dom/ReadableStreamController.h"
 #include "mozilla/dom/ReadableStreamDefaultController.h"
@@ -26,6 +27,7 @@
 class ReadableStreamDefaultReader;
 class ReadableStreamGenericReader;
 struct ReadableStreamGetReaderOptions;
+struct ReadableStreamIteratorOptions;
 struct ReadIntoRequest;
 class WritableStream;
 struct ReadableWritablePair;
@@ -140,6 +142,25 @@
                               nsTArray<RefPtr<ReadableStream>>& aResult,
                               ErrorResult& aRv);
 
+  struct IteratorData {
+    void Traverse(nsCycleCollectionTraversalCallback& cb);
+    void Unlink();
+
+    RefPtr<ReadableStreamDefaultReader> mReader;
+    bool mPreventCancel;
+  };
+
+  using Iterator = AsyncIterableIterator<ReadableStream>;
+
+  void InitAsyncIteratorData(IteratorData& aData, Iterator::IteratorType aType,
+                             const ReadableStreamIteratorOptions& aOptions,
+                             ErrorResult& aRv);
+  MOZ_CAN_RUN_SCRIPT already_AddRefed<Promise> GetNextIterationResult(
+      Iterator* aIterator, ErrorResult& aRv);
+  MOZ_CAN_RUN_SCRIPT already_AddRefed<Promise> IteratorReturn(
+      JSContext* aCx, Iterator* aIterator, JS::Handle<JS::Value> aValue,
+      ErrorResult& aRv);
+
   // Internal Slots:
  private:
   RefPtr<ReadableStreamController> mController;